Implement P3663 (Future-proof submdspan_mapping)

mhoemmen opened the pull request:
Implement P3663 (Future-proof `submdspan_mapping`).

- [X] Fix merge conflicts
- [X] Fix all check-in builds when `MDSPAN_ENABLE_P3663=OFF` (default)
- [x] Fix all check-in builds when `MDSPAN_ENABLE_P3663=ON` (this may require significant back-porting effort)
    - [x] There's probably no way to get anything resembling `constant_wrapper` in C++17 or less.  Even the C++20 "back-port" is so incomplete that it might as well be called "`integral_constant`."  We could either just use `integral_constant` for C++20 or less, or name the type `detail::constant_wrapper`.
- [x] Make `MDSPAN_ENABLE_P3663=ON` the default (retain the option so that Kokkos::View has the chance to turn it off if it affects performance)
- [X] `std::constant_wrapper`
    - [X] Remove the `constant_wrapper` back-port, as Back-port is only useful for a small range of very recent GCC and Clang implementations
    - [X] Test against GCC trunk with `-std=c++26`; we know it implements `std::constant_wrapper`, because [Compiler Explorer's version of GCC trunk has it](https://godbolt.org/z/3e4qvEvv1).
- [X] Drive-by P2389 (`dims`) fix has already been applied by PR https://github.com/kokkos/mdspan/pull/447 .  That PR will remove import of `dims` into the `experimental` namespace.  This PR should be changed not to do that.  (There's no promise of backwards compatibility for the `experimental` namespace.)
- [ ] Why is `MDSPAN_ENABLE_P3663` not defined in the code?

    - For whatever reason, CMake logic wasn't defining it in the tests.  Deal with this by defining the macro in config.h.  If we don't keep the no-P3663 branch, we can just delete the macro entirely.

- [ ] Decide whether to keep the no-P3663 branch
 
Feature is ON by default.  To disable it, set the CMake option `MDSPAN_ENABLE_P3663` to `OFF`.

Over various commits we've simplified the implementation so that it works with C++17 and subsequent C++ versions.  P3663 relies on `std::constant_wrapper`; we've tested with a version of GCC that implements it, so we can say with some confidence that it "forward-ports" all the way to C++26.

mhoemmen committed:
P3663: Add constant_wrapper implementation

It currently requires the C++23 feature "deducing this"
in order to compile.  I've tested it with dev Clang (21.0.0).
It should work fine with Clang 20.

mhoemmen committed:
P3663: Test that strided_slice works with std::cw

Also implement exposition-only __mdspan_integral_constant_like
concept, which is useful for canonicalization and Mandates.

mhoemmen committed:
P3663: Stub submdspan_canonicalize_slices

sizeof...(Slices) == 0 case builds and passes tests.

mhoemmen committed:
Support Clang 21.0.0 -std=c++2c

Set the following CMake options.

  -DCMAKE_CXX_FLAGS="-std=c++2c"

  -DMDSPAN_CXX_STANDARD=26

You'll get a warning, but it seems to work.

mhoemmen committed:
Implement check-static-bounds and others

* canonical-ice
* subtract-ice
* de-ice
* check-static-bounds

mhoemmen committed:
Fix submdspan_canonicalize_slices for one slice

mhoemmen committed:
Fix canonical_ice & 1 strided_slice

submdspan_canonicalize_slices now works for one strided_slice.
Fix canonical_ice (wrong order in subtraction; wording is fine).

mhoemmen committed:
Put dims in the correct namespace

mhoemmen committed:
Add submdspan_canonicalize_slices rank 2 test

mhoemmen committed:
CHECKPOINT

Attempt to change over submdspan to use canonicalization
without doing much else.  It fails to build; a lot of
layout mappings seem to get the wrong mapping result type.

It builds and passes tests with P3663 support off.

mhoemmen committed:
IT BUILDS -- pulled in R0 impl

mhoemmen committed:
submdspan_mapping with canonical slices actually builds now

mhoemmen committed:
CHECKPOINT

submdspan builds now when submdspan_mapping is called with
canonical slices, but the submdspan test fails.

Investigate whether this is a bug in the implementation.

mhoemmen committed:
P3663 & non-P3663 versions build & pass tests

Fix extent computation for std::complex.
It builds and passes tests.

NEXT STEPS:

* Make it ill-formed to call submdspan_mapping with non-canonical
  slice types
* Make P3663 version only use constant_wrapper and canonical
  slice types in submdspan_mapping_impl functions

mhoemmen committed:
Deploy is_canonical_slice_type in layout_left

(submdspan_mapping_impl)

mhoemmen committed:
Add is_canonical_slice_type check to all submdspan_mapping_impl

mhoemmen committed:
Check canonical k-th slice type

All submdspan_mapping_impl functions now check that every k-th slice
type is a canonical k-th slice type.  This ensures that
submdspan_mapping_impl is only well-formed for canonical k-th slice
types, and is ill-formed otherwise.

mhoemmen committed:
Replace integral_constant with cw

in P3663 branch only.

mhoemmen committed:
Remove unneeded first_of and last_of overloads

mhoemmen committed:
Remove index_pair_like definition

mhoemmen committed:
Replace is_range_slice with is_range_slice_v

mhoemmen committed:
Simplify StaticExtentFromStridedRange

Don't need to specialize for all integral-constant-like,
just for std::constant_wrapper.

mhoemmen committed:
Simplify extents_constructor for P3663

mhoemmen committed:
Remove superfluous TODO

mhoemmen committed:
Reconcile implementation with draft P3663R2

mhoemmen committed:
Add check_static_bounds test

It currently exercises the following cases for a variety
of extents types.

* full_extent_t

* integral_constant<IndexType, Value>

* IndexType

mhoemmen committed:
check_static_bounds: add convertible to full_extents_t test

mhoemmen committed:
check_static_bounds: Add strided_slice tests

mhoemmen committed:
check_static_bounds: Make test more concise

mhoemmen committed:
check_static_bounds: Improve 14.4 testing

Improve testing for the case where a structured binding of the
slice into two elements is valid.  Test includes both aggregates
and non-aggregates that opt into the tuple protocol.

mhoemmen committed:
check_static_bounds: Improve 14.4 testing more

The test should now have complete coverage of all the cases,
for all combinations of static and/or dynamic extents.

mhoemmen committed:
Add 1st draft of submdspan benchmark

Currently only the CPU version works.

mhoemmen committed:
Benchmark improvements

mhoemmen committed:
Fix convertible-to-index-type bug

The current (non-P3663) mdspan implementation incorrectly
cannot handle index or slice types that are convertible to
index_type, but are not integral-non-bool.

This commit fixes that, and adds unit tests for both
layout mapping indexing and submdspan slicing.

mhoemmen committed:
Fix P3663 benchmark and expand

Make sure that the benchmark gets the MDSPAN_ENABLE_P3663
definition (it wasn't before).

Add new benchmarks that make it harder for the compiler
to optimize away code.  (They actually change all the values
in the mdspan and check that the changes are mathematically
correct.)

In benchmark, for the slice type that's convertible to
full_extent_t, make the conversion operator noexcept.

mhoemmen committed:
Improve benchmark

CUDA version is as yet untested.

It probably won't compile, at least because the P3663 implementation
currently relies on C++23 or 26 language features that may not yet be
implemented in NVCC.

mhoemmen committed:
Refine submdspan benchmark

mhoemmen committed:
Attempt to work around GCC 11.4.0 C++20 ICE

... in constant_wrapper.  The issue is with the default argument
of the second template argument of constant_wrapper.

typename unspecified =
  typename decltype(exposition_only::cw_fixed_value(X))::type

Replacing that expression with use of the following alias doesn't help.

namespace exposition_only {
  template<auto X>
  using unspecified_t = typename decltype(cw_fixed_value(X))::type;
}

Moving the definitions of the various specializations of cw_fixed_value
above this point doesn't help either.

This generally works fine even with Clang 21.

mhoemmen committed:
Get GCC 11.4.0 (C++20) to a meaningful build error

Work around lack of parameter pack indexing
and fix constant_wrapper's assignment operator.

After some adjustments to the constant_wrapper implementation,
I found out that its overloaded arithmetic operators --
specifically, binary operator+ -- do not work.

.../submdspan_extents.hpp:96:22: error: static assertion failed
   96 |   static_assert(std::is_same_v<
      |                 ~~~~~^~~~~~~~~~
   97 |       decltype(counter + std::cw<size_t(1)>),
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   98 |       std::constant_wrapper<counter_value + size_t(1)>
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   99 |     >);

Binary operator+ doesn't kick in; the first branch of the
`is_same_v` is `unsigned long`, i.e., `size_t`.

mhoemmen committed:
Fix GCC 11.4.0 C++20 build

Define MDSPAN_CONSTANT_WRAPPER_GCC_WORKAROUND to 1
to get the work-around for GCC 11.4.0 C++20.

The key is not to rely on constant_wrapper's binary operator-.
I've tried a few designs for overloaded operators on
constant_wrapper, but none of them work with GCC 11.4.0.

mhoemmen committed:
Simplify GCC 11.4.0 C++20 work-around

mhoemmen committed:
Clean up GCC 11.4.0 C++20 work-around

Remove constant_wrapper's overloaded arithmetic operators,
since the work-around doesn't need them.

mhoemmen committed:
Fix Clang 14 C++20 build

mhoemmen committed:
Fix submdspan benchmark when P3663 is disabled

mhoemmen committed:
Make MDSPAN_CONSTANT_WRAPPER_WORKAROUND a CMake option

* Rename option for working around constant_wrapper compiler
  issues from MDSPAN_CONSTANT_WRAPPER_GCC_WORKAROUND to
  MDSPAN_CONSTANT_WRAPPER_WORKAROUND.

* Make MDSPAN_CONSTANT_WRAPPER_WORKAROUND a CMake option.
  It's OFF by default.  It only has an effect if the CMake
  option MDSPAN_ENABLE_P3663 is ON.

mhoemmen committed:
Fix GCC 15.1.0 build

GCC 15.1.0 can now build with the following options.

MDSPAN_ENABLE_P3663=ON
MDSPAN_CONSTANT_WRAPPER_WORKAROUND=OFF
CMAKE_CXX_FLAGS="-std=c++2c"
MDSPAN_CXX_STANDARD=26

mhoemmen commented:
@youyu3 Here is the branch in question.

mhoemmen committed:
Merge branch 'stable' into P3663

mhoemmen committed:
Fix build issue from merge

mhoemmen committed:
Remove extra semicolon

mhoemmen committed:
Fix some clang C++14 build errors

mhoemmen committed:
Remove C++20 requirement from benchmark

mhoemmen committed:
Fix (spurious) unused parameter warning

mhoemmen committed:
Try setting MDSPAN_ENABLE_P3663=ON by default

I expect pre-C++20 compilers not to like this.

mhoemmen committed:
Attempt to "back-port" constant_wrapper to C++17

mhoemmen committed:
Backwards-compatible integral-constant-like

Add a C++17 - compatible type trait for integral-constant-like.

mhoemmen committed:
Redefine mdspan_is_integral_constant correctly

It was only true before for std::integral_constant,
which means that the implementation wasn't conforming
to the pre-P3663 status quo.

mhoemmen committed:
Remove mdspan_is_integral_constant which is unused

mhoemmen committed:
Convert StaticExtentFromRange

so that it works for all integral-constant-like,
even when MDSPAN_ENABLE_P3663 is OFF.

mhoemmen committed:
Get rid of __mdspan_integral_constant_like

mhoemmen committed:
Introduce mdspan_constant_wrapper alias

mhoemmen committed:
Some simplifications

mhoemmen committed:
Attempt at constant_wrapper work-around

mhoemmen committed:
Some more C++17 back-porting

mhoemmen committed:
Add is_layout_mapping_alike_v

This is the C++17 back-port of the layout_mapping_alike concept.

mhoemmen committed:
Attempt to fix is_integral_constant_like_v

...and attempt to back-port some concepts to C++17.

mhoemmen committed:
Make de_ice build with C++17

Consider how we could remove the is_integral_v constraint.

mhoemmen committed:
Fix some tests

mhoemmen committed:
C++17 with P3663 ON now build and passes tests!

mhoemmen committed:
Remove C++20 lambda from submdspan definition

mhoemmen committed:
Remove some C++20 designated initializers

mhoemmen committed:
Remove unused parameter name

mhoemmen committed:
Fix more C++20 - isms

mhoemmen committed:
De-C++20-ize submdspan_canonicalize_slices

mhoemmen committed:
Fix submdspan itself for C++17, locally

mhoemmen committed:
Fix spurious unused parameter warning

mhoemmen committed:
Fix spurious unused parameter warning

mhoemmen committed:
Remove more C++20 - isms

mhoemmen committed:
Get rid of more designated initializers

mhoemmen committed:
Remove unused type alias

mhoemmen committed:
De-C++20-size test_canonicalize_slices

mhoemmen committed:
Remove C++20-isms from test_submdspan_check_static_bounds

mhoemmen committed:
Remove spurious semicolon

mhoemmen committed:
Remove more designated initializers

mhoemmen committed:
Fix some unsigned/signed comparison warnings

mhoemmen committed:
Fix more things

mhoemmen committed:
Fix more things

mhoemmen committed:
Deduplicate last_of P3663 vs. not impls

Deduplicate the P3663 ON vs. OFF versions of last_of,
by making them constrained on integral-constant-like,
instead of overloading on constant_wrapper vs. integral_constant.

mhoemmen committed:
Simplify constant_wrapper work-around

Now there are only 3 cases:

1. std::constant_wrapper and std::cw actually exist

2. We reimplement them

3. We make constant_wrapper just look like integral_constant

We now put the reimplementation in the MDSPAN_IMPL_STANDARD_NAMESPACE
namespace, instead of the std namespace.  (If (1), then we just have
using declarations that bring the two names in from the std
namespace.)

mhoemmen committed:
Fix some (un)signed comparison warnings

mhoemmen committed:
Fix unused parameter warning

mhoemmen committed:
Work around icpx bug

mhoemmen committed:
Hide is_layout_mapping_alike from C++ < 17

as it depends on std::bool_constant, a C++17 feature.

mhoemmen commented on include/experimental/__p2630_bits/strided_slice.hpp:
```diff
@@ -40,9 +59,14 @@ struct strided_slice {
+  static_assert(detail::__mdspan_is_index_like_v<StrideType>);
 };
 
+#if (__cplusplus < 202002L)
```
Consider removing this deduction guide.

mhoemmen replied:
I've removed `strided_slice`'s deduction guide.

mhoemmen commented on include/experimental/__p2630_bits/strided_slice.hpp:
```diff
@@ -17,17 +17,36 @@
+
+#if defined(MDSPAN_ENABLE_P3663)
+  template<class T, T Value>
+  using mdspan_constant_wrapper = decltype(cw<Value>);
```
Consider removing `mdspan_constant_wrapper`.

mhoemmen replied:
I removed it.

mhoemmen committed:
PREVIOUS COMMIT PASSES ALL CHECK-IN TESTS

Small cleanup

mhoemmen committed:
DRIVE-BY FIX: Actually implement P2389

dims was in the experimental namespace;
it belongs in the main namespace.

mhoemmen commented on include/experimental/__p2630_bits/strided_slice.hpp:
```diff
@@ -40,9 +53,14 @@ struct strided_slice {
+  static_assert(detail::__mdspan_is_index_like_v<StrideType>);
 };
 
+#if (__cplusplus < 202002L)
```
Consider using mdspan's macros for C++ version instead of `__cplusplus`.

mhoemmen replied:
I don't use this macro any more.

mhoemmen commented on include/experimental/__p2630_bits/submdspan_extents.hpp:
```diff
@@ -17,6 +17,7 @@
 #pragma once
 
 #include <complex>
+#include <tuple>
```
Why do we need this?

mhoemmen replied:
The header should have included `<tuple>` before, because there were specializations of `index_pair_like` for `std::tuple`.

mhoemmen commented on tests/test_dims.cpp:
```diff
@@ -31,7 +31,7 @@ static constexpr bool is_extents_v<
 void test_dims_with_one_template_argument()
 {
-  using d = MDSPAN_IMPL_STANDARD_NAMESPACE :: MDSPAN_IMPL_PROPOSED_NAMESPACE :: dims<Rank>;
+  using d = MDSPAN_IMPL_STANDARD_NAMESPACE :: dims<Rank>;
```
This is part of the P2389 drive-by fix.

mhoemmen replied:
I've fixed merge conflicts with `stable` as a result of merging PR #447 .

mhoemmen commented on include/experimental/__p2630_bits/submdspan_extents.hpp:
```diff
@@ -85,28 +128,106 @@ struct index_pair_like<std::array<IdxT, 2>, IndexType> {
+)
+MDSPAN_INLINE_FUNCTION
+constexpr Integral first_of(const Integral &i) {
+  // FIXME (mfh 2025/06/06) This is broken, but it's better than it was.
```
We should just fix the non-P3663 branch if we intend to keep it.

mhoemmen commented on include/experimental/__p2630_bits/submdspan_extents.hpp:
```diff
@@ -85,28 +128,106 @@ struct index_pair_like<std::array<IdxT, 2>, IndexType> {
+  return {};
+}
+#else
+// NOTE This is technically not conforming.
```
We should just fix this.

mhoemmen commented on include/experimental/__p2630_bits/submdspan_extents.hpp:
```diff
@@ -150,49 +273,105 @@ first_of(const strided_slice<OffsetType, ExtentType, StrideType> &r) {
+// clang++ with C++14 is not fond of the pragma appearing inside the
+// macro definition.  In that case, it complains, "error: embedding a
+// directive within macro arguments has undefined behavior
+// [-Werror,-Wembedded-directive]."  The fix is to duplicate code.
```
This describes code that was later removed.  I think we can delete this comment.

mhoemmen replied:
I've deleted this comment.

mhoemmen commented on include/experimental/__p2630_bits/submdspan_extents.hpp:
```diff
@@ -268,6 +479,21 @@ constexpr auto divide(const T0 &v0, const T1 &v1) {
+#if defined(MDSPAN_ENABLE_P3663)
+template <class IndexType, auto v0, auto v1>
+MDSPAN_INLINE_FUNCTION
+constexpr auto divide(constant_wrapper<v0> i0,
```
Very few compilers actually support `constant_wrapper` arithmetic yet.  That's why we have these functions.

(EDIT: There's nothing to fix here; I'm just pointing this out to reviewers.)

mhoemmen commented on include/experimental/__p2630_bits/submdspan_extents.hpp:
```diff
@@ -404,6 +681,571 @@ struct extents_constructor<0, Extents, NewStaticExtents...> {
+  constexpr check_static_bounds_result check_static_bounds(
+    const extents<IndexType, Exts...>&)
+{
+#if defined(__cpp_pack_indexing)
```
It might make sense just to remove the branch and define a function for getting the k-th type of a pack.  (See `get_kth_in_pack` below.)

mhoemmen replied:
I've done this.

mhoemmen commented on include/experimental/__p2630_bits/submdspan_extents.hpp:
```diff
@@ -404,6 +681,571 @@ struct extents_constructor<0, Extents, NewStaticExtents...> {
+// in constant expressions if they are integral-constant-like.
+
+template<size_t k, class S_k, class IndexType, size_t... Exts>
+  constexpr check_static_bounds_result check_static_bounds(
```
FIXME In R3, this returns `bool` instead of an enum.

mhoemmen commented on include/experimental/__p2630_bits/submdspan_extents.hpp:
```diff
@@ -404,6 +681,571 @@ struct extents_constructor<0, Extents, NewStaticExtents...> {
+  check_canonical_kth_subdmspan_slice_types_impl(
+    std::make_index_sequence<sizeof...(Slices)>(), exts, slices...);
+#else
+  // We really want to keep the C++20 branch here
```
Consider removing this branch.  It's harder to maintain two implementations.  Kokkos requires C++20 or greater, so the pre-C++20 branch wouldn't be tested as often if it's kept separate.

mhoemmen replied:
I've removed this branch.

mhoemmen commented on include/experimental/__p2630_bits/submdspan_extents.hpp:
```diff
@@ -404,6 +681,571 @@ struct extents_constructor<0, Extents, NewStaticExtents...> {
+  }
+  else if constexpr (is_strided_slice<Slice>::value) { // 11.3
+#if ! defined(MDSPAN_CONSTANT_WRAPPER_WORKAROUND)
+    // GCC 11.4.0 (C++20) accepts this code, but Clang 14 does not.
```
Just delete this branch.  icpx also doesn't like this (and its feature test macros don't prevent warnings-as-errors).

mhoemmen replied:
I've removed this branch.

mhoemmen commented on include/experimental/__p2630_bits/submdspan_extents.hpp:
```diff
@@ -404,6 +681,571 @@ struct extents_constructor<0, Extents, NewStaticExtents...> {
+  else if constexpr (detail::is_std_complex<Slice>) {
+#if ! defined(MDSPAN_CONSTANT_WRAPPER_WORKAROUND)
+    // GCC 11.4.0 (C++20) accepts this code, but Clang 14 does not.
+    return strided_slice{
```
Just delete this branch.

mhoemmen replied:
I did : - )

mhoemmen commented on include/experimental/__p2630_bits/submdspan_extents.hpp:
```diff
@@ -404,6 +681,571 @@ struct extents_constructor<0, Extents, NewStaticExtents...> {
+    static_assert(std::is_convertible_v<S_k1, IndexType>);
+#if ! defined(MDSPAN_CONSTANT_WRAPPER_WORKAROUND)
+    // GCC 11.4.0 (C++20) accepts this code, but Clang 14 does not.
+    return strided_slice{
```
Just delete this branch.

mhoemmen replied:
I've removed this branch.

mhoemmen commented on include/experimental/__p2630_bits/submdspan_extents.hpp:
```diff
@@ -404,6 +681,571 @@ struct extents_constructor<0, Extents, NewStaticExtents...> {
+  const extents<IndexType, Extents...>& exts,
+  Slices... slices)
+{
+  return std::tuple{
```
Consider using `Kokkos::tuple` instead of `std::tuple`.

mhoemmen replied:
This requires replacing use of `std::apply` in `submdspan`.

mhoemmen commented on include/experimental/__p2630_bits/submdspan_mapping.hpp:
```diff
@@ -60,10 +139,42 @@ namespace detail {
+  return static_cast<common_t>(first_of(slice)) ==
          static_cast<common_t>(ext);
+#else
+  // NOTE (mfh 2025/06/06) The original implementation was not conforming.
```
We should decide whether to fix the non-P3663 case or just remove it and make P3663 the only implementation.

mhoemmen commented on tests/test_strided_slice.cpp:
```diff
@@ -0,0 +1,103 @@
+template<class OffsetType, class ExtentType, class StrideType>
+void test_strided_slice(OffsetType offset, ExtentType extent, StrideType stride)
+{
+  // Some compilers are bad at CTAD for aggregates.
```
We should consider branching based on compilers where it's known to work.  It would be good to test that CTAD works.

nmm0 commented on benchmarks/submdspan/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+
+#if(MDSPAN_ENABLE_CUDA)
+#  add_subdirectory(cuda)
+#endif()
```
```suggestion
```

mhoemmen replied:
Fixed; thanks! : - )

nmm0 commented on include/experimental/__p2630_bits/constant_wrapper.hpp:
```diff
@@ -0,0 +1,265 @@
+#pragma once
```
I think we should pull this into a separate PR.

Also since we are pulling this in should we ask Zach if it's under a license or public domain? I couldn't find any info in the source code so I suspect it is the latter.

mhoemmen replied:
@nmm0 I'm inclined to check whether any (VERY VERY NEW) compiler comes with a `constant_wrapper` implementation.  If it does, we can test the code with the actual `std::constant_wrapper`, remove the back-port, and leave only the pre-C++26 work-around (that looks like `integral_constant`).  What do you think?

mhoemmen replied:
Compiler Explorer's GCC trunk has `std::constant_wrapper`: https://godbolt.org/z/3e4qvEvv1 .  That means we have a way to test this with the actual `std::constant_wrapper`.  That motivates removing the back-port, as long as we know that it actually works with `std::constant_wrapper`.

C++26 may still see changes to `std::constant_wrapper` -- there's a fix paper in flight -- but I don't think it would apply to anything in this PR.  (The paper would change behavior for callables.)

nmm0 replied:
It does look like GCC16 will have `std::constant_wrapper` (so i guess it made the cut): https://gcc.gnu.org/gcc-16/changes.html

I'm fine with having a constant wrapper backport though, just will be easier to review in another PR

mhoemmen replied:
@nmm0 The back-port is really only useful for Clang 21 and just barely for GCC 15.  Anything earlier won't have a chance of building a `constant_wrapper` that looks enough like `constant_wrapper` to call it a back-port.

I've removed the back-port.  GCC trunk builds it correctly.  I'm not convinced that `MDSPAN_ENABLE_P3663` is always defined in the build, though.

mhoemmen replied:
I've unconditionally defined the macro `MDSPAN_ENABLE_P3663` in config.h (see above) and GCC trunk builds everything.

mhoemmen commented:
PR https://github.com/kokkos/mdspan/pull/447 will remove import of `dims` into the `experimental` namespace.  This PR should be changed not to do that.  (There's no promise of backwards compatibility for the `experimental` namespace.)

nmm0 commented on include/experimental/__p2630_bits/remove_cvref.hpp:
```diff
@@ -0,0 +1,20 @@
+
+#include <type_traits>
+
+namespace MDSPAN_IMPL_STANDARD_NAMESPACE {
```
I think this is fine here but at some point if we reorganize files I think we could move all of these little backports of C++ things that we need into its own utility folder.

mhoemmen replied:
That's a good idea!  If we end up removing the `constant_wrapper` back-port (please see my comment above), then there won't be quite so many long files here.

mhoemmen replied:
I ended up removing the `constant_wrapper` back-port, btw.

nmm0 commented on include/experimental/__p2630_bits/strided_slice.hpp:
```diff
@@ -17,17 +17,30 @@
-  struct mdspan_is_integral_constant<std::integral_constant<T,val>>: std::true_type {};
-}
+  template<class T>
+  constexpr bool __mdspan_is_index_like_v =
```
Just trying to wrap my brain around this -- is this the same behavior as before?

mhoemmen replied:
@nmm0 It's not the same behavior as before.  The existing implementation is NOT conforming.  Even before P3663, _`integral-constant-like`_ is a concept that user-defined types can meet.  However, the code only called `std::integral_constant` and `Kokkos::integral_constant` "integral constant like."

mhoemmen replied:
I've fixed `__mdspan_is_index_like_v` so that it's now conforming even in the no-P3663 case.  There are other places in which the no-P3663 case is still nonconforming, though.  The comments in the code should point those out.

nmm0 replied:
ohhh I see. Thanks for fixing it

nmm0 commented on include/experimental/__p2630_bits/strided_slice.hpp:
```diff
@@ -40,9 +53,14 @@ struct strided_slice {
 
+#if (__cplusplus < 202002L)
+template <class OffsetType, class ExtentType, class StrideType>
+strided_slice(const OffsetType&, const ExtentType&, const StrideType&) ->
```
Is this deduction guide needed? I know you said to consider removing it but it seems like this shouldn't do anything unless I'm missing something

mhoemmen replied:
It didn't actually help with icpx (which is a Clang derivative if I remember right).  We should remove it; aggregates like this shouldn't need a deduction guide.  For broken compilers, users would have to work around for other things anyway.

mhoemmen replied:
I've removed the deduction guide for `strided_slice`.

nmm0 commented on include/experimental/__p2630_bits/submdspan.hpp:
```diff
@@ -20,12 +20,58 @@
+
+#if defined(MDSPAN_ENABLE_P3663)
+
+#  if defined(__cpp_structured_bindings) && (__cpp_structured_bindings >= 202411L)
```
To be honest unless this really improves compile times I'd just stick with the tuple implementation so we only have one implementation to maintain/understand

mhoemmen replied:
I've removed the C++26 branch.

nmm0 commented on include/experimental/__p2630_bits/submdspan.hpp:
```diff
@@ -20,12 +20,58 @@
+    submdspan_canonicalize_slices(src.extents(), slices...);
+  using src_mapping_type = decltype(src.mapping()); // CTAD doesn't seem to work
+  auto sub_map_result = std::apply(
+    detail::submdspan_mapping_caller<src_mapping_type>{src.mapping()},
```
This might be a little more readable as a lambda and we wouldn't need the template argument since mapping would be captured. This way we don't have to jump around to see the definition

mhoemmen replied:
I've reimplemented this as a lambda with `auto&&... canonical_slices` capture.  Thanks!

nmm0 commented on include/experimental/__p2630_bits/submdspan_extents.hpp:
```diff
@@ -85,28 +128,106 @@ struct index_pair_like<std::array<IdxT, 2>, IndexType> {
 MDSPAN_TEMPLATE_REQUIRES(
   class Integral,
-  /* requires */(std::is_convertible_v<Integral, size_t>)
+  /* requires */(std::is_signed_v<Integral> || std::is_unsigned_v<Integral>)
```
Does this actually need any constraints specifying it is integral? Or it would never get to the point where its called by a float or something

mhoemmen replied:
P3663 means that `first_of` only ever sees

1. `full_extent_t`,
2. `index_type` or _`integral-constant-like`_ of `index_type`, or
3. `strided_slice` with template parameters satisfying (2).

I've simplified this a bit to make that more clear, and to remove the need for `requires` clauses on `first_of` in the P3663 case.  (I just use function overloads.)

nmm0 commented on CMakeLists.txt:
```diff
@@ -28,10 +28,15 @@ option(MDSPAN_GENERATE_STD_NAMESPACE_TARGETS "Whether to generate and install ta
 
 option(MDSPAN_ENABLE_CONCEPTS "Try to enable concepts support by giving extra flags." On)
 
+option(MDSPAN_ENABLE_P3663 "Enable implementation of P3663 (Future-proof submdspan_mapping)." On)
```
I think this is actually our only cmake-based feature switch? So this actually leads to the question of how we do this, because right now if someone is using mdspan via cmake this does nothing (as it only enables a compile definition for benchmarks).

So I suppose the question is do we want to add them to a target (e.g. via target_compile_definitions) or via a `#cmakedefine` sort of thing like Kokkos does where we generate a header with the define. We could even let it be overrideable via a define.

I actually am leaning towards the latter because we could use it in our single-header mode and the value could be overriden if using the header only mode via a define.

mhoemmen replied:
@nmm0 wrote:
> ... because right now if someone is using mdspan via cmake this does nothing....

Practically, it changes the implementation by changing the macro definitions in the headers -- at least it should!  I might be missing something here.

Should we consider just removing this option and making the library implement P3663 unconditionally?

nmm0 reviewed (commented):
Ok some initial comments, I got through a good chunk of it but not everything.

mhoemmen committed:
Remove strided_slice deduction guide

mhoemmen commented:
Compiler Explorer's version of GCC trunk has `std::constant_wrapper`: https://godbolt.org/z/3e4qvEvv1 .

mhoemmen committed:
Merge branch 'stable' into P3663

mhoemmen committed:
Fix dims merge

mhoemmen committed:
Fix __mdspan_is_index_like_v

For both P3663 and no-P3663 branches, fix __mdspan_is_index_like_v
so that it now correctly excludes std::bool_constant or any
integral-constant-like with bool value type.

mhoemmen committed:
Remove C++26 branch from submdspan

mhoemmen committed:
Simplify submdspan per review comment

to use a lambda with `auto&&... canonical_slices` capture
instead of a named functor.

mhoemmen committed:
Simplify first_of definition

mhoemmen committed:
Simplify first_of more

mhoemmen committed:
Remove outdated comment

mhoemmen committed:
Fixes so that GCC trunk builds

GCC trunk ("16.0.1") has std::constant_wrapper.
This commit fixes any build errors that resulted from
using that as the "constant_wrapper" in mdspan.
It also works around some pack indexing issues
that seem unique to GCC.

mhoemmen commented:
Hi @nmm0 !  GCC trunk ("16.0.1") has `std::constant_wrapper`.  Commit 971c8fb6068bcebf4399cf4a18e8133633c0ec91 builds and passes tests with that, using the following cmake options.

```
cmake ${SOURCE_DIRECTORY} \
  -DMDSPAN_ENABLE_TESTS=ON \
  -DMDSPAN_ENABLE_EXAMPLES=ON \
  -DMDSPAN_ENABLE_P3663=ON \
  -DCMAKE_CXX_FLAGS="-Wall -std=c++26" \
  -DCMAKE_CXX_COMPILER=${GCC_TRUNK_PREFIX}/bin/g++ \
  -DMDSPAN_CXX_STANDARD=26
```

As a result, I think it's safe to remove the `constant_wrapper` back-port entirely.

mhoemmen committed:
Remove constant_wrapper back-port

GCC trunk ("16.0.1") has std::constant_wrapper.
This commit builds and passes tests with that.
Thus, I've removed the constant_wrapper back-port
(that really only worked with GCC 15 and Clang 21).
There are now only two branches.

1. Use actual std::constant_wrapper and std::cw

2. Use something that only superficially has the same
   interface but is really just an integral_constant clone

mhoemmen committed:
Fix inv_map_rank so it uses constant_wrapper

...instead of integral_constant.

Add FIXME comments for later, where the definition doesn't quite
match up with the Standard (either pre- or post-P3663).

mhoemmen committed:
Unconditionally define MDSPAN_ENABLE_P3663

mhoemmen commented on benchmarks/submdspan/cuda/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+
+if(CMAKE_CUDA_COMPILER_ID STREQUAL "NVIDIA")
+  set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} --extended-lambda")
```
We need to test with this flag removed.

mhoemmen committed:
Remove commented-out CMake logic

mhoemmen commented on include/experimental/__p0009_bits/config.hpp:
```diff
@@ -303,3 +303,7 @@ static_assert(MDSPAN_IMPL_CPLUSPLUS >= MDSPAN_CXX_STD_14, "mdspan requires C++14
 #endif
+
+#if ! defined(MDSPAN_ENABLE_P3663)
+#  define MDSPAN_ENABLE_P3663 1
```
We may want to revisit this if we decide we want to keep the no-P3663 branch.

mhoemmen committed:
A bit of cleanup

mhoemmen committed:
Add FIXME comments and reformat code a bit

mhoemmen committed:
Add missing 'typename'
